tests/test_operator_overloading.py

```diff
@@ -126,3 +126,18 @@ def test_nested():
     assert abase.value == 42
     del abase, b
     pytest.gc_collect()
+
+
+def test_overriding_eq_reset_hash():
+
+    assert m.Comparable(15) is not m.Comparable(15)
+    assert m.Comparable(15) == m.Comparable(15)
+
+    with pytest.raises(TypeError):
+        hash(m.Comparable(15))  # TypeError: unhashable type: 'm.Comparable'
+
+    for hashable in (m.Hashable, m.Hashable2):
+        assert hashable(15) is not hashable(15)
+        assert hashable(15) == hashable(15)
+
+        assert hash(hashable(15)) == hash(hashable(15))
```

Comment:
Is this to prove that the correct hash implementation is used?
Does `hash(hashable(15)) == hash(hashable(15 + 32))` make sense?
To prove that the `% 32` is actually used.

Reply:
If the default `id`-based comparison would be used, this assertion would fail (I believe? Unless the memory location of the temporary `hashable(15)` gets recycled?

Reply:
Yeah, I agree, but it is so cheap to be more certain, I'd definitely seize the opportunity.

Reply:
The `__hash__` contract implies that equal object has same hash. This is what is checked  here (along with hashibility of the objects). 

>  Does hash(hashable(15)) == hash(hashable(15 + 32)) make sense?

Only as verification of concrete implementation. The exact hash values are implementation detail. 

@YannickJadoul, you are correct, default hash would fail here.

Reply:
> @YannickJadoul, you are correct, default hash would fail here.

Did you try this? Because I could imagine Python creating a `hashable(15)` at memory location `x`, computing the `hash`, deallocating the `hashable` object as it goes to refcount 0, the creating `hashable(15)` at the same memory location, and computing the same `id`-based hash?

Reply:
Yes, I've tried to mock it with 
```python
        assert hashable(15) is not hashable(15)
        assert hashable(15) == hashable(15)
        del hashable.__hash__
        assert hash(hashable(15)) == hash(hashable(15))
```
and it fails with 

```
>           assert hash(hashable(15)) == hash(hashable(15))
E           AssertionError: assert -9223363257492360940 == 8779362415047
E            +  where -9223363257492360940 = hash(<pybind11_tests.operators.Hashable object at 0x7fc1ac731148>)
E            +    where <pybind11_tests.operators.Hashable object at 0x7fc1ac731148> = <class 'pybind11_tests.operators.Hashable'>(15)
E            +  and   8779362415047 = hash(<pybind11_tests.operators.Hashable object at 0x7fc1ac731c70>)
E            +    where <pybind11_tests.operators.Hashable object at 0x7fc1ac731c70> = <class 'pybind11_tests.operators.Hashable'>(15)
```

Reply:
I actually just tried that. I knew I'd seen this Python trap before:

```pycon
>>> class X: pass
... 
>>> id(X()), id(X())
(140256801227776, 140256801227776)
>>> hash(X()) == hash(X())
True
>>> x = X()
>>> y = X()
>>> id(x), id(y)
(140256801227776, 140256801227832)
>>> hash(x) == hash(y)
False
```

The same holds for running as a script:

```python
class X: pass

print(id(X()), id(X()))
print(hash(X()) == hash(X()))

x, y = X(), X()

print(id(x), id(y))
print(hash(x) == hash(y))
```

```
$ python3 test.py 
140000331939064 140000331939064
True
140000331939064 140000331959096
False
```

Reply:
Hm. I don't mind to test against actual values of `hash(hashable(...))` to verify that we use the definition we want.

Reply:
I've added `assert hash(hashable(15)) == 15` so we know for sure the default `id`-based implementation is not involved.

Reply:
OK, that's also fine. I was just thinking to not have them as temporary objects, but this does the same, I suppose.

Reply:
> Yes, I've tried to mock it with
> 
> ```python
>         assert hashable(15) is not hashable(15)
>         assert hashable(15) == hashable(15)
>         del hashable.__hash__
>         print(hashable.__hash__)
>         assert hash(hashable(15)) == hash(hashable(15))
> ```
> 
> and it fails with
> 
> ```
> <slot wrapper '__hash__' of 'object' objects>
> 
> ...
> 
> >           assert hash(hashable(15)) == hash(hashable(15))
> E           AssertionError: assert -9223363257492360940 == 8779362415047
> E            +  where -9223363257492360940 = hash(<pybind11_tests.operators.Hashable object at 0x7fc1ac731148>)
> E            +    where <pybind11_tests.operators.Hashable object at 0x7fc1ac731148> = <class 'pybind11_tests.operators.Hashable'>(15)
> E            +  and   8779362415047 = hash(<pybind11_tests.operators.Hashable object at 0x7fc1ac731c70>)
> E            +    where <pybind11_tests.operators.Hashable object at 0x7fc1ac731c70> = <class 'pybind11_tests.operators.Hashable'>(15)
> ```

Ah, wait. Hadn't seen that. Interesting, so here it doesn't do that? Maybe because of the AST rewrite pytest does?

Reply:
Whatever it is I don't think it's a reliable behavior, i.e. it might change from version to version from one day of week to another. I think now we are backed with better test case.

Reply:
Agreed! :-)